### ISSUE TITLE
derivable_impls triggers on dyn Trait, where using a concrete type for default

### ISSUE THREAD

**Issue by nihohit** [rust-lang/rust-clippy#13032]
### Summary

The `derivable_impls` warning triggers if all fields of a struct are set using a default initializer, even if the fields are declared as dyn traits that can't be derived, and the default initializer is called using an implementing struct.

### Lint Name

derivable_impls

### Reproducer

I tried this code:

```rust
#![allow(dead_code)]

use std::sync::Arc;
trait Foo {
}

#[derive(Default)]
struct Impl {}

impl Foo for Impl {}

struct Container(Arc<dyn Foo>);

impl Default for Container {
    fn default() -> Self {
        Self(Arc::<Impl>::default())
    }
}
```

I saw this happen:

```
warning: this `impl` can be derived
  --> src/lib.rs:14:1
   |
14 | / impl Default for Container {
15 | |     fn default() -> Self {
16 | |         Self(Arc::<Impl>::default())
17 | |     }
18 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#derivable_impls
   = note: `#[warn(clippy::derivable_impls)]` on by default
   = help: remove the manual implementation...
help: ...and instead derive it
   |
12 + #[derive(Default)]
13 | struct Container(Arc<dyn Foo>);
   |

warning: `playground` (lib) generated 1 warning (run `cargo clippy --fix --lib -p playground` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.45s
```

I expected to see this happen:
no warning

### Version

```text
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: aarch64-apple-darwin
release: 1.79.0
LLVM version: 18.1.7
```
